orderfilter/validate_js.go

```diff
@@ -0,0 +1,130 @@
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	orderValidatorLoaded   = false
+	messageValidatorLoaded = false
+	// validatorLoadCheckInterval is frequently to check whether the schema
+	// validators have been loaded.
+	validatorLoadCheckInterval = 50 * time.Millisecond
+)
+
+type SchemaValidationResult struct {
+	valid  bool
+	errors []error
+}
+
+func (s *SchemaValidationResult) Valid() bool {
+	return s.valid
+}
+
+func (s *SchemaValidationResult) Errors() []error {
+	return s.errors
+}
+
+// func checkForOrderValidator() bool {
+// 	if js.Global().Get("schemaValidator").Get("orderValidator") != js.Undefined() && js.Global().Get("orderValidator") != js.Null() {
+// 		log.Info("Order Validator finished loading")
+// 		orderValidatorLoaded = true
+// 		return true
+// 	}
+// 	return false
+// }
+//
+// func checkForMessageValidator() bool {
+// 	if js.Global().Get("schemaValidator").Get("messageValidator") != js.Undefined() && js.Global().Get("messageValidator") != js.Null() {
+// 		log.Info("Message Validator finished loading")
+// 		messageValidatorLoaded = true
+// 		return true
+// 	}
+// 	return false
+// }
+
+// ValidateOrderJSON Validates a JSON encoded signed order using the AJV javascript library.
+// This libarary is used to increase the performance of Mesh nodes that run in the browser.
+func (f *Filter) ValidateOrderJSON(orderJSON []byte) (*SchemaValidationResult, error) {
+	//	if !orderValidatorLoaded {
+	//		loaded := checkForOrderValidator()
+	//		if !loaded {
+	//			return nil, errors.New("order validator not loaded")
+	//		}
+	//	}
+
+	jsResult := js.Global().Get("schemaValidator").Call("orderValidator", js.ValueOf(string(orderJSON)))
+	fatal := jsResult.Get("fatal")
+	if !jsutil.IsNullOrUndefined(fatal) {
+		return nil, errors.New(fatal.String())
+	}
+	valid := jsResult.Get("success").Bool()
+	jsErrors := jsResult.Get("errors")
+	var convertedErrors []error
+	for i := 0; i < jsErrors.Length(); i++ {
+		convertedErrors = append(convertedErrors, errors.New(jsErrors.Get(fmt.Sprintf("%d", i)).String()))
+	}
+	return &SchemaValidationResult{valid: valid, errors: convertedErrors}, nil
+}
+
+// MatchOrder returns true if the order passes the filter. It only returns an
+// error if there was a problem with validation. For details about
+// orders that do not pass the filter, use ValidateOrder.
+func (f *Filter) MatchOrder(order *zeroex.SignedOrder) (bool, error) {
+	result, err := f.ValidateOrder(order)
+	if err != nil {
+		return false, err
+	}
+	return result.Valid(), nil
+}
+
+func (f *Filter) MatchOrderMessageJSON(messageJSON []byte) (bool, error) {
+	//	if !messageValidatorLoaded {
+	//		loaded := checkForMessageValidator()
+	//		if !loaded {
+	//			return false, errors.New("message validator not loaded")
+	//		}
+	//	}
+
+	jsResult := js.Global().Get("schemaValidator").Call("messageValidator", js.ValueOf(string(messageJSON)))
```

Comment:
If `ajv` accepts `Uint8Array` or `Buffer`, it might be more efficient to use [`CopyBytesToGo`](https://golang.org/pkg/syscall/js/#CopyBytesToGo) and/or [`CopyBytesToJS`](https://golang.org/pkg/syscall/js/#CopyBytesToJS) instead of casting to/from `string`. Worth checking.

Reply:
From what I can tell, it doesn't. I'll experiment with this a little bit more (in a simpler example), and see what happens.

Reply:
No worries, don't spend anymore time on it if we already have something fast enough.

Reply:
Sounds good. I'll probably try it out this weekend, but unless I read the `ajv` source incorrectly (it's pretty opaque, so that's completely possible), it validates the data "as-is" without any other processing.